tests/returntypes/test_spsa_gradient_new.py

```diff
@@ -0,0 +1,1052 @@
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        f0 = dev.execute(tape)
+        n = 9
+        tapes, fn = spsa_grad(tape, strategy="forward", approx_order=1, num_directions=n, f0=f0)
+        # one tape per direction, the unshifted one already was evaluated above
+        assert len(tapes) == n
+
+    def test_independent_parameters(self):
+        """Test the case where expectation values are independent of some parameters. For those
+        parameters, the gradient should be evaluated to zero without executing the device."""
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(1.0, wires=[0])
+            qml.RX(1.0, wires=[1])
+            qml.expval(qml.PauliZ(0))
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RX(1.0, wires=[0])
+            qml.RX(1.0, wires=[1])
+            qml.expval(qml.PauliZ(1))
+
+        n1 = 5
+        tapes, fn = spsa_grad(tape1, approx_order=1, strategy="forward", num_directions=n1)
+        j1 = fn(dev.batch_execute(tapes))
+
+        assert len(tapes) == dev.num_executions == n1 + 1
+
+        n2 = 11
+        tapes, fn = spsa_grad(tape2, num_directions=n2)
+        j2 = fn(dev.batch_execute(tapes))
+
+        assert len(tapes) == 2 * n2
+
+        # Because there is just a single gate parameter varied for these tapes, the gradient
+        # approximation will actually be as good as finite differences.
+        exp = -np.sin(1)
+
+        assert np.allclose(j1, [exp, 0])
+        assert np.allclose(j2, [0, exp])
+
+    def test_output_shape_matches_qnode(self):
+        """Test that the transform output shape matches that of the QNode."""
+        dev = qml.device("default.qubit", wires=4)
+
+        def cost1(x):
+            qml.Rot(*x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        def cost2(x):
+            qml.Rot(*x, wires=0)
+            return [qml.expval(qml.PauliZ(0))]
+
+        def cost3(x):
+            qml.Rot(*x, wires=0)
+            return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+        def cost4(x):
+            qml.Rot(*x, wires=0)
+            return qml.probs([0, 1])
+
+        def cost5(x):
+            qml.Rot(*x, wires=0)
+            return [qml.probs([0, 1])]
+
+        def cost6(x):
+            qml.Rot(*x, wires=0)
+            return qml.probs([0, 1]), qml.probs([2, 3])
+
+        x = np.random.rand(3)
+        circuits = [qml.QNode(cost, dev) for cost in (cost1, cost2, cost3, cost4, cost5, cost6)]
+
+        transform = [qml.math.shape(spsa_grad(c)(x)) for c in circuits]
+
+        expected = [(3,), (3,), (2, 3), (3, 4), (3, 4), (2, 3, 4)]
+
+        assert all(t == q for t, q in zip(transform, expected))
+
+    # TODO: added with param shift
```

Comment:
We should be able to add this one.

Reply:
Yes, it will work regarding the support of special observables with SPSA. However, this test uses the `QNode` kwarg `diff_method`, which is not registered yet. I uncommented this test in all three test files and marked it with `pytest.mark.skip`. In #3440 the registration is added, and we will be able to add the tests correspondingly :)

Reply:
Ok thank you!